server/NewBankClientHandler.java

```diff
@@ -23,46 +29,64 @@ public void run() {
-						if (customer != null) {
-							out.println("\nLoading...\n");
-							out.println("Log In Successful.\n");
-							out.println("Welcome to your NewBank account " + customer.getUsername()+".");
-							out.println("\n+---------------------------------------------------+");
-							out.println("|NewBank - Customer Menu                            |");
-							out.println("+---------------------------------------------------+");
-							out.println("|(1) - Show existing accounts and current balance   |");
-							out.println("|(2) - Create a new account with zero opening funds |");
-							out.println("|(3) - Move funds between your existing accounts    |");
-							out.println("|(4) - Pay an existing NewBank customer             |");
-							out.println("+---------------------------------------------------+");
-							while (true) {
-								String request = in.readLine();
-								System.out.println("Request from " + customer.getUsername());
-								// Added pass newBankClientHandler Object into processRequest, so it can call methods
-								String response = bank.processRequest(customer, request, this);
-								out.println(response);
+
+						//If 3 failed login attempts during session, does not allow login
+						if (failedLogInCount >2) {
+							out.println("There have been too many unsuccessful login attempts, " +
+									"please disconnect and try again later.");
+							break;
+						}
+
+						while (failedLogInCount<3) {
+							// Ask for username
+							out.println("Enter Username:");
+							String userName = in.readLine();
+
+							/*
+							checks if account is locked, note will only be useful once database is running and
+							the database stores accountLocked variable. Currently resetting connection resets
+							the customer HashMap
+							*/
+							if (isAccountLocked(userName)){
+								out.println("This account is locked.");
+								break;
+							}
+							// Ask for password
+							out.println("Enter Password:");
+							String password = in.readLine();
+							out.println("\nChecking Details...");
+
+							// Authenticate user and get customer ID token from bank for use in subsequent requests.
+							Customer customer = bank.checkLogInDetails(userName, password, this);
+
+							// If the user is authenticated then get requests from the user and process them.
+							if (customer != null) {
+								showCustomerMenu(customer);
+								while (true) {
+									String request = in.readLine();
+									System.out.println("Request from " + customer.getUsername());
+									// Added pass newBankClientHandler Object into processRequest, so it can call methods
+									String response = bank.processRequest(customer, request, this);
+									out.println(response);
+								}
+							} else {
+								out.println("Loading...");
+								out.println("Log In Failed.\n");
+								//increases failed log in count by one and adds the string of username to failed login list.
+								failedLogInCount += 1;
```

Comment:
The way you implemented this wouldn't really work as desired from my understanding. `failedLogInCount` tracks the total number of failed logins, not failed logins per user, so for example, I first try to login as Matija, fail 2 times, then I try to login as Stephen, fail once, and that's it, I can't login anymore, I have to disconnect and try again.

You could implement something like `countMap` in `userLogInFrequency`, and store usernames and number of failed logins, and then decide depending on that.

Reply:
I think my reasoning was to have two different mechanisms, one for total login attempts and one for username locking.
This is to prevent a user from attempting as many login attempts into different accounts as they liked with say '1234' and hoping to crack one. So we could set the failedLogInCount as 5, but still have an account lock after 3 attempts. 
It still does lock the user if you key in the same name 3 times.

I have made some changes:
1. I have removed the failedLogInCount as failedLogInUsers.size() does the same thing. I have set the total login attempts to 5 now so you can keep trying to login after locking one account.
2. userLogInFrequency() now returns the countMap as I wanted to keep the method to doing one thing
3. added lockUsers which takes that countMap, and for every username, it checks the frequency and if the account is not yet locked, locks it and displays a message.